component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java

```diff
@@ -0,0 +1,262 @@
+                                Status.DEADLINE_EXCEEDED));
+                        streamObserverMap.remove(messageId);
+                    }
+                } finally {
+                    readWriteLock.writeLock().unlock();
+                }
+            }
+        }
+    }
+
+    @Override
+    public void initSource(OptionHolder optionHolder, String[] requestedTransportPropertyNames) {
+        this.sourceId = optionHolder.validateAndGetOption(GrpcConstants.SOURCE_ID).getValue();
+        this.requestedTransportPropertyNames = requestedTransportPropertyNames.clone();
+        this.serviceTimeout = Long.parseLong(optionHolder.getOrCreateOption(GrpcConstants.SERVICE_TIMEOUT,
+                GrpcConstants.SERVICE_TIMEOUT_DEFAULT).getValue());
+        this.timer = new Timer();
+        GrpcSourceRegistry.getInstance().putGrpcServiceSource(sourceId, this);
+    }
+
+    @Override
+    public void connect(ConnectionCallback connectionCallback, State state) throws ConnectionUnavailableException {
+        try {
+            server.start();
+            if (logger.isDebugEnabled()) {
+                logger.debug(siddhiAppContext.getName() + ": Server started");
+            }
+        } catch (IOException e) {
+            throw new SiddhiAppRuntimeException(siddhiAppContext.getName() + ": " + e.getMessage());
+        }
+    }
+
+    /**
+     * This method can be called when it is needed to disconnect from the end point.
+     */
+    @Override
+    public void disconnect() {
+        try {
+            Server serverPointer = server;
+            if (serverPointer == null) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(siddhiAppContext.getName() + ": Illegal state. Server already stopped.");
+                }
+                return;
+            }
+            serverPointer.shutdown();
+            if (serverPointer.awaitTermination(serverShutdownWaitingTime, TimeUnit.SECONDS)) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(siddhiAppContext.getName() + ": Server stopped");
+                }
+                return;
+            }
+            serverPointer.shutdownNow();
+            if (serverPointer.awaitTermination(serverShutdownWaitingTime, TimeUnit.SECONDS)) {
+                return;
+            }
+            throw new SiddhiAppRuntimeException(siddhiAppContext.getName() + ": Unable to shutdown server");
+        } catch (InterruptedException e) {
+            throw new SiddhiAppRuntimeException(siddhiAppContext.getName() + ": " + e.getMessage());
+        }
+    }
+
+    @Override
+    public void populateHeaderString(String headerString) {
+        this.headerString = headerString;
+    }
+
+    public void handleCallback(String messageId, String responsePayload) {
+        if (isDefaultMode) {
+            readWriteLock.writeLock().lock();
+            try {
+                StreamObserver<Event> streamObserver = streamObserverMap.get(messageId);
```

Comment:
Cant we do a sync map and only do remove here that will return the removed StreamObserver